xds/src/main/java/io/grpc/xds/XdsPicker.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.LoadBalancer.PickResult;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.LoadBalancer.SubchannelPicker;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+final class XdsPicker<LocalityT> extends SubchannelPicker {
+
+  private final List<LocalityT> localities;
+  private final List<Integer> weights;
+  private final Map<LocalityT, SubchannelPicker> childPickers;
+  private final Random random;
+  private final int totalWeight;
+
+  XdsPicker(
+      List<LocalityT> localities,
+      List<Integer> weights,
+      Map<LocalityT, SubchannelPicker> childPickers) {
+    this(localities, weights, childPickers, new Random());
+  }
+
+  @VisibleForTesting
+  XdsPicker(
+      List<LocalityT> localities,
+      List<Integer> weights,
+      Map<LocalityT, SubchannelPicker> childPickers,
+      Random random) {
+
+    checkNotNull(localities, "localities in null");
+    checkArgument(
+        !localities.isEmpty(),
+        "localities list is empty");
+    this.localities = ImmutableList.copyOf(localities);
+
+    checkNotNull(weights, "weights in null");
+    checkArgument(
+        weights.size() == localities.size(),
+        "localities and weights are of different sizes");
```

Comment:
It may be simpler to define a `WeightedLocality` class:
```java
class WeightedLocality {
  int getWeight();
  SubchannelPicker getPicker();
}
```
Then you wouldn't need most of the checks here and you wouldn't need the map either.